File: src/contest/other/Cyclopian_Puzzle.java
package contest.other;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Cyclopian_Puzzle {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

	public static void main (String[] args) throws NumberFormatException, IOException {
		solve(Integer.parseInt(br.readLine()), 'A', 'B', 'C');
	}

	public static void solve (int n, char origin, char temp, char dest) {
		if (n == 0)
			return;

		solve(n - 1, origin, dest, temp);
		System.out.println(origin + "" + dest);
		solve(n - 1, temp, origin, dest);
	}
}

File: src/contest/misc/Cyclopian_Puzzle.cc
#include <bits/stdc++.h>

using namespace std;
static void solve(int n, char start, char temp, char dest) {
  if (n == 0)
    return;
  solve(n - 1, start, dest, temp);
  printf("%c%c\n", start, dest);
  solve(n - 1, temp, start, dest);
}
int main() {
  int n;
  scanf("%d", &n);
  solve(n, 'A', 'B', 'C');
  return 0;
}
